### ISSUE TITLE
[Feature Proposal] Conditional Batch Norm

### ISSUE THREAD

**Issue by t-vi** [pytorch/pytorch#8985]
I'm implementing conditional batch norm.
If we (you) want it, I'd be glad to send a PR.

**Comment by bhushan23** [comment]
this sounds interesting, can you give some stats the benefits?

**Comment by t-vi** [comment]
Ah, sorry, I should have added "...as featured by conditional [SN-GAN](https://github.com/pfnet-research/sngan_projection)". The idea is to have the output stats (weight and bias) but not input (running_var/mean) depend on a class parameter.

**Comment by Kaixhin** [comment]
Is this just batch norm/instance norm with `affine=False` and using an `Embedding` to produce a weight and bias? If so then using the existing components would be simple enough, but I guess it is popular enough that a premade component would be nice.

**Comment by t-vi** [comment]
@Kaixhin yup. That is exactly it.

**Comment by t-vi** [comment]
So here is the code I'm using, copypasting much of BatchNorm, a bit of that could also be achieved by subclassing.
```
class CategoricalConditionalBatchNorm(torch.nn.Module):
    """
    Similar to batch norm, but with per-category weight and bias.
    """
    def __init__(self, num_features, num_cats, eps=2e-5, momentum=0.1, affine=True,
                 track_running_stats=True):
        super().__init__()
        self.num_features = num_features
        self.num_cats = num_cats
        self.eps = eps
        self.momentum = momentum
        self.affine = affine
        self.track_running_stats = track_running_stats
        if self.affine:
            self.weight = torch.nn.Parameter(torch.Tensor(num_cats, num_features))
            self.bias = torch.nn.Parameter(torch.Tensor(num_cats, num_features))
        else:
            self.register_parameter('weight', None)
            self.register_parameter('bias', None)
        if self.track_running_stats:
            self.register_buffer('running_mean', torch.zeros(num_features))
            self.register_buffer('running_var', torch.ones(num_features))
            self.register_buffer('num_batches_tracked', torch.tensor(0, dtype=torch.long))
        else:
            self.register_parameter('running_mean', None)
            self.register_parameter('running_var', None)
            self.register_parameter('num_batches_tracked', None)
        self.reset_parameters()

    def reset_running_stats(self):
        if self.track_running_stats:
            self.running_mean.zero_()
            self.running_var.fill_(1)
            self.num_batches_tracked.zero_()

    def reset_parameters(self):
        self.reset_running_stats()
        if self.affine:
            self.weight.data.fill_(1.0)
            self.bias.data.zero_()

    def forward(self, input, cats):
        exponential_average_factor = 0.0

        if self.training and self.track_running_stats:
            self.num_batches_tracked += 1
            if self.momentum is None:  # use cumulative moving average
                exponential_average_factor = 1.0 / self.num_batches_tracked.item()
            else:  # use exponential moving average
                exponential_average_factor = self.momentum

        out = torch.nn.functional.batch_norm(
            input, self.running_mean, self.running_var, None, None,
            self.training or not self.track_running_stats,
            exponential_average_factor, self.eps)
        if self.affine:
            shape = [input.size(0), self.num_features] + (input.dim() - 2) * [1]
            weight = self.weight.index_select(0, cats).view(shape)
            bias = self.bias.index_select(0, cats).view(shape)
            out = out * weight + bias
        return out

    def extra_repr(self):
        return '{num_features}, num_cats={num_cats}, eps={eps}, momentum={momentum}, affine={affine}, ' \
               'track_running_stats={track_running_stats}'.format(**self.__dict__)
```

**Comment by Kaixhin** [comment]
Here's the code I'm using, think it's better to reduce code reuse and just use the components we have already:

```py
class ConditionalBatchNorm2d(nn.Module):
  def __init__(self, num_features, num_classes):
    super().__init__()
    self.num_features = num_features
    self.bn = nn.BatchNorm2d(num_features, affine=False)
    self.embed = nn.Embedding(num_classes, num_features * 2)
    self.embed.weight.data[:, :num_features].normal_(1, 0.02)  # Initialise scale at N(1, 0.02)
    self.embed.weight.data[:, num_features:].zero_()  # Initialise bias at 0

  def forward(self, x, y):
    out = self.bn(x)
    gamma, beta = self.embed(y).chunk(2, 1)
    out = gamma.view(-1, self.num_features, 1, 1) * out + beta.view(-1, self.num_features, 1, 1)
    return out
```

**Comment by bhushan23** [comment]
[BIG-GAN](https://arxiv.org/abs/1809.11096) is also using Conditional Batch Norm and has achieved highest Inception Score. 

I believe this is worth adding.

**Comment by Hshihua** [comment]
Ah, Can anyone tell me the difference between the Categorical Conditional BN and Conditional BN? In my opinion, there is no difference, just different name.

**Comment by askerlee** [comment]
Another option of conditional batchnorm implemented by me, which doesn't share running_mean and running_var across domains:
`
class CondBatchNorm2d(nn.Module):
    def __init__(self, num_features, eps=1e-05, momentum=0.05, 
                 affine=True, track_running_stats=True, num_domains=2):
        super().__init__()
        self.num_features = num_features
        self.bns = [] # nn.ModuleList()
        self.num_domains = num_domains
        for i in range(self.num_domains):
            self.bns.append(nn.BatchNorm2d(num_features, eps, momentum, affine, track_running_stats))
            self.bns[i].weight.data.fill_(1)
            self.bns[i].bias.data.zero_()

    def _load_from_state_dict(self, state_dict, prefix, metadata, strict, 
                            missing_keys, unexpected_keys, error_msgs):
        # if bns are not in a ModuleList yet, then we are loading a cp of the traditional model, 
        # so we manually copy the same weights/bias/running_mean/running_var/num_batches_tracked to all bns
        # if bns are in a ModuleList, it means we are loading a checkpoint of a multidomain model.
        # bns not need to be manually initialized. pytorch will initialize ModuleList, which in turn initializes bns
        if type(self.bns) == list:        
            for i in range(self.num_domains):
                self.bns[i]._load_from_state_dict( \
                            state_dict, prefix, metadata, strict, 
                            missing_keys, unexpected_keys, error_msgs )
    
    # default is in the first domain. 
    # if domains is not specified, fall back to the original bn's behavior
    def forward(self, x, domains=None):
        # convert from list to ModuleList, so that torch.save() will save parameters of bns
        # Couldn't do the conversion in _load_from_state_dict; otherwise pytorch will 
        # directly access the elements in bns and try to load them, 
        # which will cause "missing keys" errors
        if type(self.bns) == list:
            self.bns = nn.ModuleList(self.bns)
            if x.is_cuda:
                self.bns.cuda()
                
        if domains is None:
            domains = torch.zeros_like(x[:, 0, 0, 0]).long()
        out = torch.zeros_like(x)
        for i in range(self.num_domains):
            if (domains == i).sum() > 0:
                x_dom = x[domains==i]
                out[domains==i] = self.bns[i](x_dom)
        return out

`

**Comment by askerlee** [comment]
Another option of conditional batchnorm implemented by me, which doesn't share running_mean and running_var across domains:

``` 
class CondBatchNorm2d(nn.Module):
    def __init__(self, num_features, eps=1e-05, momentum=0.05, 
                 affine=True, track_running_stats=True, num_domains=2):
        super().__init__()
        self.num_features = num_features
        self.bns = [] # nn.ModuleList()
        self.num_domains = num_domains
        for i in range(self.num_domains):
            self.bns.append(nn.BatchNorm2d(num_features, eps, momentum, affine, track_running_stats))
            self.bns[i].weight.data.fill_(1)
            self.bns[i].bias.data.zero_()

    def _load_from_state_dict(self, state_dict, prefix, metadata, strict, 
                            missing_keys, unexpected_keys, error_msgs):
        # if bns are not in a ModuleList yet, then we are loading a cp of the traditional model, 
        # so we manually copy the same weights/bias/running_mean/running_var/num_batches_tracked to all bns
        # if bns are in a ModuleList, it means we are loading a checkpoint of a multidomain model.
        # bns not need to be manually initialized. pytorch will initialize ModuleList, which in turn initializes bns
        if type(self.bns) == list:        
            for i in range(self.num_domains):
                self.bns[i]._load_from_state_dict( \
                            state_dict, prefix, metadata, strict, 
                            missing_keys, unexpected_keys, error_msgs )
    
    # default is in the first domain. 
    # if domains is not specified, fall back to the original bn's behavior
    def forward(self, x, domains=None):
        # convert from list to ModuleList, so that torch.save() will save parameters of bns
        # Couldn't do the conversion in _load_from_state_dict; otherwise pytorch will 
        # directly access the elements in bns and try to load them, 
        # which will cause "missing keys" errors
        if type(self.bns) == list:
            self.bns = nn.ModuleList(self.bns)
            if x.is_cuda:
                self.bns.cuda()
                
        if domains is None:
            domains = torch.zeros_like(x[:, 0, 0, 0]).long()
        out = torch.zeros_like(x)
        for i in range(self.num_domains):
            if (domains == i).sum() > 0:
                x_dom = x[domains==i]
                out[domains==i] = self.bns[i](x_dom)
        return out
```

**Comment by JohnnyRisk** [comment]
> Here's the code I'm using, think it's better to reduce code reuse and just use the components we have already:
> 
> ```python
> class ConditionalBatchNorm2d(nn.Module):
>   def __init__(self, num_features, num_classes):
>     super().__init__()
>     self.num_features = num_features
>     self.bn = nn.BatchNorm2d(num_features, affine=False)
>     self.embed = nn.Embedding(num_classes, num_features * 2)
>     self.embed.weight.data[:, :num_features].normal_(1, 0.02)  # Initialise scale at N(1, 0.02)
>     self.embed.weight.data[:, num_features:].zero_()  # Initialise bias at 0
> 
>   def forward(self, x, y):
>     out = self.bn(x)
>     gamma, beta = self.embed(y).chunk(2, 1)
>     out = gamma.view(-1, self.num_features, 1, 1) * out + beta.view(-1, self.num_features, 1, 1)
>     return out
> ```

![batch_norm_equation](https://user-images.githubusercontent.com/18605542/55833263-8906c680-5ad4-11e9-92a0-46d8fbca432d.png)

Sorry If I have misunderstood your code but, in self.bn(x) doesn't this already have a saved gamma and beta which it applies to the input? 

According to the equation above which comes from the Modulating Early Visual Processing by Language paper (https://papers.nips.cc/paper/7237-modulating-early-visual-processing-by-language.pdf) we would then have Gamma_Embedding * ( original bn) + Beta_Embedding 

The line I am referring to is (out = gamma.view(-1, self.num_features, 1, 1) * out + beta.view(-1, self.num_features, 1, 1))
I think means that we would be multiplying the original gamma and beta by the embedding gamma. Is this what we want to do? 

As far as I can tell the implementation from @t-vi does not do this but, please let me know if I am mistaken.

**Comment by ezyang** [comment]
@soumith can you take a look at this?

**Comment by soumith** [comment]
@t-vi if you are still interested in a PR, I'd be down. It looks like a simple python-level addition.
Would like to hear form @ajbrock and @Kaixhin on whether there's an API that generalizes across applications or we'd have to think hard about it. Just above, there are two variants, but there's not much disagreement.

**Comment by t-vi** [comment]
So one thing that borders on bikeshedding: with StyleGAN's Linear layer for weights the ConditionalBatchNorm here is really a CategoricalConditionalBatchNorm (I think the SNGAN uses the term somewhere). That wasn't as clear to me a year ago.

**Comment by Kaixhin** [comment]
Yes after a batch norm (without an affine transformation) it's then a question of whether you want an affine transformation for continuous values (use `nn.Linear`) or discrete values (use `nn.Embedding`). I'm not sure it's worth the bloat for having two different functions/modules that do this, but one could be worth it (accepting discrete values by default).

**Comment by ajbrock** [comment]
I agree with Kai that `nn.Linear` and `nn.Embedding` will cover most use cases, and it feels like it would be worth having a module that supports either. I think the most general API would be one that just takes in `(x, gain, bias)` and returns `gain * normalize(x) +bias)` (potentially as a wrapper around an arbitrary normalizer so that it can use InstanceNorm, GroupNorm, etc, instead of just BN) but that puts much more burden on the user to handle computing and resizing the gains and biases themselves, which an advanced or tinkering user is just going to do on their own anyway. 

One other note: BigGAN uses 1-centered gains (`out = (1 + gain) * normalize(x) + bias`) because the gains come from linear layers, and it's more complex to initialize those to output 1s than it is with `nn.Embedding` (which you can just set to be ~1). If you want to include the linear layer in the ConditionalBN module, this is probably worth considering.

**Comment by t-vi** [comment]
Clearly this is obsolete after 3 years, it won't happen.
